File: app/src/main/java/com/example/purepawapp/ui/product/ProductCardAdapter.java
package com.example.purepawapp.ui.product;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.data.model.Product;
import com.example.purepawapp.databinding.ItemProductCardBinding;
import com.example.purepawapp.util.CurrencyUtils;

import java.util.function.Consumer;

public class ProductCardAdapter extends ListAdapter<Product, ProductCardAdapter.ViewHolder> {

    private final Consumer<Product> onClick;
    private final Consumer<Product> onAddToCart;

    private static final DiffUtil.ItemCallback<Product> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.equals(newItem);
        }
    };

    public ProductCardAdapter(Consumer<Product> onClick, Consumer<Product> onAddToCart) {
        super(DIFF_CALLBACK);
        this.onClick = onClick;
        this.onAddToCart = onAddToCart;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemProductCardBinding binding = ItemProductCardBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemProductCardBinding binding;

        ViewHolder(ItemProductCardBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Product product) {
            ProductUi.loadImage(binding.ivProductImage, product);
            binding.tvBadge.setText(ProductUi.getBadgeLabel(product));
            binding.tvName.setText(product.getName());
            binding.tvPrice.setText(CurrencyUtils.toVndString(product.getDisplayPrice()));
            binding.getRoot().setOnClickListener(v -> onClick.accept(product));
            binding.btnAddToCart.setOnClickListener(v -> onAddToCart.accept(product));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/product/ProductCardAdapter.kt
package com.example.purepawapp.ui.product

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.data.model.Product
import com.example.purepawapp.databinding.ItemProductCardBinding
import com.example.purepawapp.util.toVndString

class ProductCardAdapter(
    private val onClick: (Product) -> Unit,
    private val onAddToCart: (Product) -> Unit
) : ListAdapter<Product, ProductCardAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemProductCardBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemProductCardBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(product: Product) {
            binding.tvIcon.text = product.emoji
            binding.tvBadge.text = product.badgeLabel
            binding.tvName.text = product.name
            binding.tvPrice.text = product.displayPrice.toVndString()
            binding.root.setOnClickListener { onClick(product) }
            binding.btnAddToCart.setOnClickListener { onAddToCart(product) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Product>() {
            override fun areItemsTheSame(oldItem: Product, newItem: Product): Boolean = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: Product, newItem: Product): Boolean = oldItem == newItem
        }
    }
}
